portfolio/src/main/java/com/google/sps/data/CalendarGroup.java

```diff
@@ -0,0 +1,126 @@
+   *     working hour start time and the current timestamp.
+   * @param endTime: end of possible scheduling blocks, of type Instant.
+   */
+  public CalendarGroup(
+      List<CalendarEvent> events, Instant overallStartTime, Instant overallEndTime) {
+    if (events == null) {
+      throw new IllegalArgumentException("Calendar Group needs a collection of events");
+    }
+    if (overallStartTime == null) {
+      throw new IllegalArgumentException("Calendar Group needs the start of scheduling hours");
+    }
+    if (overallEndTime == null) {
+      throw new IllegalArgumentException("Calendar Group needs the end of scheduling hours");
+    }
+    if (overallStartTime.isAfter(overallEndTime)) {
+      throw new IllegalArgumentException("Start time cannot be after end time");
+    }
+    this.events = events;
+    this.overallStartTime = overallStartTime;
+    this.overallEndTime = overallEndTime;
+    this.allFreeTimeRanges = calculateFreeTimeRanges();
+  }
+
+  /**
+   * Calculates a TimeRange list which represents the periods of time that are empty of events and
+   * lie completely inside the possible scheduling hours. Originally written by tomasalvarez,
+   * modified by hollyyuqizheng.
+   */
+  public List<TimeRange> calculateFreeTimeRanges() {
+    Collections.sort(events, sortByEventStartTimeAscending);
+
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // This represents the earliest time that we can schedule a window for the
+    // meeting. As events are processed, this changes to their end times.
+    Instant earliestNonScheduledInstant = overallStartTime;
+
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and it is not
+      // later than the person's scheduling hours' ending time.
+      if (event.getStartTime().isAfter(earliestNonScheduledInstant)
+          && !event.getStartTime().isAfter(overallEndTime)) {
+        possibleTimes.add(
+            TimeRange.fromStartEnd(earliestNonScheduledInstant, event.getStartTime()));
+      }
+      // Check if the earliest non scheduled time needs to be shifted to later.
+      if (earliestNonScheduledInstant.isBefore(event.getEndTime())) {
+        earliestNonScheduledInstant = event.getEndTime();
+      }
+    }
+    // The end of the work hours is potentially never included so we check.
+    if (overallEndTime.isAfter(earliestNonScheduledInstant)) {
+      possibleTimes.add(TimeRange.fromStartEnd(earliestNonScheduledInstant, overallEndTime));
+    }
+    return possibleTimes;
+  }
+
+  /** Getter for the list of all free time ranges. */
+  public List<TimeRange> getFreeTimeRanges() {
+    return allFreeTimeRanges;
+  }
+
+  /**
+   * Deletes a time range from the list of all free times.
+   *
+   * @return the modifed list of free time ranges.
+   */
+  public List<TimeRange> deleteFreeTimeRange(TimeRange timeRange) {
```

Comment:
This method and the one below seem a bit weird to me, because I think free time ranges are computed based on events and overall start/end instants. I think it would make more sense to expose methods for adding/removing calendar events. And then at any point the caller can get the list of free time ranges. So consider removing these time range modification methods. And I don't think they are called anywhere except for in tests.

Reply:
I see. I was thinking of handling more of the free time range in this class, but I can imagine that all the free time range can be handled in a different class or in the same class as the scheduling algorithm? For right now, I removed these two methods and added a method for adding individual calendar event to the group, and I don't think a method for removing events will be helpful in this case.

Reply:
Okay sounds good